prov/efa/src/efa_domain.c

```diff
@@ -88,18 +88,25 @@ static struct fi_ops_domain efa_ops_domain_rdm = {
  *
  * @param efa_domain[in,out]	efa domain to be set.
  * @param domain_name		domain name
+ * @param ep_type		endpoint type
  * @return 0 if efa_domain->device and efa_domain->ibv_pd has been set successfully
  *         negative error code if err is encountered
  */
-static int efa_domain_init_device_and_pd(struct efa_domain *efa_domain, const char *domain_name)
+static int efa_domain_init_device_and_pd(struct efa_domain *efa_domain,
+                                         const char *domain_name,
+                                         enum fi_ep_type ep_type)
 {
 	int i;
+	char* device_name = NULL;
+	const char* domain_name_suffix = efa_domain_name_suffix(ep_type);
 
 	if (!domain_name)
 		return -FI_EINVAL;
 
 	for (i = 0; i < g_device_cnt; i++) {
-		if (strstr(domain_name, g_device_list[i].ibv_ctx->device->name) == domain_name) {
+		device_name = g_device_list[i].ibv_ctx->device->name;
+		if (strstr(domain_name, device_name) == domain_name &&
+		    strlen(domain_name) - strlen(device_name) == strlen(domain_name_suffix)) {
```

Comment:
can we do an exact match (using strcmp) between domain_name and device_name + domain_name_suffix?

Reply:
I considered that option, but my understanding is that approach requires an extra memory allocation to construct device_name + domain_name_suffix.

From a correctness perspective I think the current approach is also reasonable, because we know domain_name = device_name + domain_name_suffix, and the `if` test verifies 2 things:
1. The domain name is prefixed by device name
2. The domain name length is the same as device name + suffix

So the only corner case would be `efa_10-rdm` is matched against `efa_1` when requested ep_type is DGRAM, i.e. DGRAM and RDM are used together, which should not happen at all.

I'm open to do the full string match if you think the tradeoff is worthwhile.

Reply:
The simpler solution is to `memcmp()` up to `domain_name_suffix[0]` (i.e. $n_{bytes} =$ `(ptrdiff_t) memchr(domain_name, *domain_name_suffix) - domain_name`. No extra null-terminated string required.